src/serialize.rs

```diff
@@ -0,0 +1,992 @@
+            },
+            ErrorKind::WrongArrayLen { actual, expected } => {
+                write!(f, "wrong array size (expected {}, got {})", expected, actual)
+            },
+            ErrorKind::WrongFields { actual, expected } => {
+                write!(f, "field names do not match (expected {:?}, got {:?})", expected, actual)
+            },
+            ErrorKind::BadScalar { type_str, rust_type, verb } => {
+                write!(f, "cannot {} type {} with type-string '{}'", verb, rust_type, type_str)
+            },
+            ErrorKind::UsizeOverflow(value) => {
+                write!(f, "cannot cast {} as usize", value)
+            },
+        }
+    }
+}
+
+impl<T> TypeRead for Box<dyn TypeRead<Value=T>> {
+    type Value = T;
+
+    #[inline(always)]
+    fn read_one<'a>(&self, bytes: &'a [u8]) -> (T, &'a [u8]) {
+        (**self).read_one(bytes)
+    }
+}
+
+impl<T: ?Sized> TypeWrite for Box<dyn TypeWriteDyn<Value=T>> {
+    type Value = T;
+
+    #[inline(always)]
+    fn write_one<W: io::Write>(&self, mut writer: W, value: &T) -> io::Result<()>
+    where Self: Sized,
+    {
+        // Boxes must always go through two virtual dispatches.
+        //
+        // (one on the TypeWrite trait object, and one on the Writer which must be
+        //  cast to the monomorphic type `&mut dyn io::write`)
+        (**self).write_one_dyn(&mut writer, value)
+    }
+}
+
+fn invalid_data<T>(message: &str) -> io::Result<T> {
+    Err(io::Error::new(io::ErrorKind::InvalidData, message.to_string()))
+}
+
+// Takes info about each data size, from largest to smallest.
+macro_rules! impl_integer_serializable {
+    ( @iterate
+        meta: $meta:tt
+        remaining: []
+    ) => {};
+
+    ( @iterate
+        meta: $meta:tt
+        remaining: [$first:tt $($smaller:tt)*]
+    ) => {
+        impl_integer_serializable! {
+          @generate
+            meta: $meta
+            current: $first
+        }
+
+        impl_integer_serializable! {
+          @iterate
+            meta: $meta
+            remaining: [ $($smaller)* ]
+        }
+    };
+
+    (
+        @generate
+        meta: [ (main_ty: $Int:ident) (date_ty: $DateTime:ident) ]
+        current: [ $size:literal $int:ident
+                   (size1: $size1_cfg:meta) $read_int:ident $write_int:ident
+                 ]
+    ) => {
+        mod $int {
+            use super::*;
+
+            pub struct AnyEndianReader { pub(super) swap_byteorder: bool }
+            pub struct AnyEndianWriter { pub(super) swap_byteorder: bool }
+
+            pub(super) fn expect_scalar_dtype(dtype: &DType) -> Result<&TypeStr, DTypeError> {
+                dtype.as_scalar().ok_or_else(|| {
+                    DTypeError::expected_scalar(dtype, stringify!($int))
+                })
+            }
+
+            #[inline]
+            fn maybe_swap(swap: bool, x: $int) -> $int {
+                match swap {
+                    true => x.to_be().to_le(),
```

Comment:
Isn't `to_be().to_le()` just `swap_bytes()`?

```suggestion
                    true => x.swap_bytes(),
```

Reply:
Oops, didn't know that existed!